### PR TITLE
txn: Forbid alloc space every time for `OptimisticTxnContextProvider`

### LINKED ISSUE #36986: Optimize `NewOptimisticTxnContextProvider` to reduce memory allocations.
## Enhancement

In sysbench test we found `NewOptimisticTxnContextProvider` will always be called and introduces some unnecessary memory allocations:

image

Maybe we can optimize that to reduce the above cost

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #36986

Problem Summary:

### Check List

Tests 

- [x] Manual test (add detailed scripts or steps below)

Use sysbench to check oltp_point_get in my local macbook with union store:

```
sysbench oltp_point_select run --mysql-host=127.0.0.1 --mysql-port=4001 --mysql-user=root \
--db-driver=mysql --mysql-db=test --threads=50 --time=120 --report-interval=10 \
--tables=32 --table-size=1000 \
--mysql-ignore-errors=1062,2013,8028,9007 --auto-inc=false
```

master:
```
SQL statistics:
    queries performed:
        read:                            8607193
        write:                           0
        other:                           0
        total:                           8607193
    transactions:                        8607193 (71720.96 per sec.)
    queries:                             8607193 (71720.96 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

Throughput:
    events/s (eps):                      71720.9603
    time elapsed:                        120.0095s
    total number of events:              8607193

Latency (ms):
         min:                                    0.02
         avg:                                    0.70
         max:                                  462.90
         95th percentile:                        0.00
         sum:                              5999189.60

Threads fairness:
    events (avg/stddev):           172143.8600/2794.94
    execution time (avg/stddev):   119.9838/0.00
```

image

This PR:

```
SQL statistics:
    queries performed:
        read:                            8676243
        write:                           0
        other:                           0
        total:                           8676243
    transactions:                        8676243 (72298.29 per sec.)
    queries:                             8676243 (72298.29 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

Throughput:
    events/s (eps):                      72298.2920
    time elapsed:                        120.0062s
    total number of events:              8676243

Latency (ms):
         min:                                    0.02
         avg:                                    0.69
         max:                                  368.71
         95th percentile:                        0.00
         sum:                              5999021.04

Threads fairness:
    events (avg/stddev):           173524.8600/1773.51
    execution time (avg/stddev):   119.9804/0.00
```

image

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by tiancaiamao — APPROVED** [review]

File: session/txnmanager.go [hunk 1]
```diff
@@ -49,6 +49,9 @@ type txnManager struct {
 
 	stmtNode    ast.StmtNode
 	ctxProvider sessiontxn.TxnContextProvider
+
+	// We always reuse a same OptimisticTxnContextProvider in one session to reduce memory allocation cost for every new txn
```
Comment (tangenta) [inline session/txnmanager.go:53]:
```suggestion
	// We always reuse the same OptimisticTxnContextProvider in one session to reduce memory allocation cost for every new txn.
```

File: sessiontxn/isolation/optimistic.go [hunk 2]
```diff
@@ -27,28 +27,27 @@ import (
...
-	}
+// ResetForNewTxn resets OptimisticTxnContextProvider to an initial state for a new txn
+func (p *OptimisticTxnContextProvider) ResetForNewTxn(sctx sessionctx.Context, causalConsistencyOnly bool) {
+	*p = emptyOptimisticTxnContextProvider
+	p.sctx = sctx
+	p.causalConsistencyOnly = causalConsistencyOnly
+	p.onTxnActiveFunc = p.onTxnActive
+	p.getStmtReadTSFunc = p.getTxnStartTS
+	p.getStmtForUpdateTSFunc = p.getTxnStartTS
+}
```
Comment (tangenta) [inline sessiontxn/isolation/optimistic.go:46]: What about `p.optimizeWithMaxTS`?

Comment (lcwangchao) [reply sessiontxn/isolation/optimistic.go:46]: optimizeWithMaxTS should be false by default. `*p = emptyOptimisticTxnContextProvider` reseted all values to default.

**Review by tangenta — APPROVED** [review]

**Pull request merged**